### PR TITLE
Change to having single source for realm prefixes

### PR DESCRIPTION
**Pull request by backspace**
Right now `@cardstack/` realm prefix setup is handled differently in `realm-server` and `host`, this consolidates that.

### CODE REVIEW

File: packages/host/config/environment.js [hunk 1]
```diff
@@ -225,10 +225,23 @@ module.exports = function (environment) {
...
+    // realm prefixes. Nulling it unconditionally here made a host built for
+    // tests disagree with a realm-server that had started the catalog, and
+    // `internalKeyFor` resolves through each process's own prefix set — so the
+    // definitions cache one writes is keyed differently from what the other
+    // reads. Host CI sets SKIP_CATALOG=true, so this is already undefined
+    // there; what changes is that it is no longer undefined when the catalog
+    // realm is in fact running.
+    if (skipCatalog) {
+      ENV.resolvedCatalogRealmURL = undefined;
+    }
```
[the remark this answers is not in the document]
Comment (backspace) [reply packages/host/config/environment.js]: [Claude Code 🤖] Correct, and fixed in 907003c — thanks. The rewrite is now explicitly non-test.

You identified the consequence I checked only halfway: I confirmed host CI wouldn't break (it sets `SKIP_CATALOG=true`, so `resolvedCatalogRealmURL` stays undefined there) without checking what happens when the new branch is actually taken.

Verified by evaluating the config rather than reasoning about it:

| case | `resolvedCatalogRealmURL` | default ids |
| -- | -- | -- |
| test, `SKIP_CATALOG` unset | `https://localhost:4201/catalog/` | `http://test-realm/test/...` |
| test, `SKIP_CATALOG=true` | undefined | `http://test-realm/test/...` |
| development | `https://localhost:4201/catalog/` | catalog-derived |

So the catalog prefix mapping is kept — which is the point of the change — while the default ids stay test-owned, and they now match `main` for every test configuration. Development is unaffected.

Comment (habdelra) [inline packages/host/config/environment.js]: [Claude Code 🤖] Regression, blocking-worthy. `defaultSystemCardId` was not the only thing the unconditional null was holding back. Evaluating this file for `test` with `SKIP_CATALOG` unset — the local default — now yields `resolvedCatalogRealmURL = https://localhost:4201/catalog/`, and three more sites read it:

- `packages/host/app/lib/utils.ts` — `catalogRealmURL` goes from `null` to that URL, which flips the `catalogRealmURL &&` branch that builds a realm entry in `tests/acceptance/ai-assistant-test.gts`.
- `tests/helpers/realm-server-mock/routes.ts` — the `/_catalog-realms` handler returns the real catalog realm alongside skills, for every test that does not call `setCatalogRealmURL`.
- `tests/helpers/mock-matrix/_client.ts` — `publicRealmURLs` gains it at module load.

None of this is visible to CI, which sets `SKIP_CATALOG=true`, so a green run says nothing about the branch being added here. What ships is a local host-test configuration that quietly differs from CI's.

Two ways out. Either keep `resolvedCatalogRealmURL` nulled in the test environment and carry the catalog's URL in a separate config key that only `PREFIX_REALMS` reads, so widening the prefix set does not widen anything else; or leave it as written and say, per site, why the new behaviour is the intended one. The second owes an answer for openrouter too — it stays unconditionally nulled on a test-isolation argument that applied just as well to the catalog.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/host/config/environment.js [hunk 2]
```diff
@@ -225,10 +225,23 @@ module.exports = function (environment) {
...
-    // cost-tier lookup point this at their own test realm explicitly.
+    // The catalog realm follows SKIP_CATALOG, the same signal the realm-server
+    // gates START_CATALOG on, so the two processes register the same set of
+    // realm prefixes. Nulling it unconditionally here made a host built for
+    // tests disagree with a realm-server that had started the catalog, and
+    // `internalKeyFor` resolves through each process's own prefix set — so the
+    // definitions cache one writes is keyed differently from what the other
+    // reads. Host CI sets SKIP_CATALOG=true, so this is already undefined
+    // there; what changes is that it is no longer undefined when the catalog
+    // realm is in fact running.
```
Comment (habdelra) [inline packages/host/config/environment.js]: [Claude Code 🤖] Pre-existing, now load-bearing; non-blocking. "the same signal the realm-server gates `START_CATALOG` on, so the two processes register the same set" holds only for `SKIP_CATALOG=true` or unset. There are two spellings of the variable in the tree:

- `mise-tasks/services/realm-server` — `[ "${SKIP_CATALOG:-}" = "true" ]`, matching this file's `=== 'true'`
- `mise-tasks/services/worker` — `[ -z "${SKIP_CATALOG:-}" ]`, so any non-empty value skips

`SKIP_CATALOG=1` therefore starts the catalog realm on the realm-server, registers `@cardstack/catalog/` on the host, and leaves the worker without it — and the worker is the process that reads back the definitions cache the host's prerender writes, which is the divergence this declaration exists to close. Normalising the worker to `= "true"` is a one-line change and makes the claim in this comment true; otherwise the comment should say which process it means.

Separately, both new comment blocks in this file describe the change rather than the contract — "Nulling it unconditionally here made a host built for tests disagree", "what changes is that it is no longer undefined", and at line 255 "The test environment previously reached this block with `resolvedCatalogRealmURL` always undefined, so the guard was implicit; now that…". A reader arriving later has no anchor for when "now" was. Stating the invariants directly covers it: the catalog realm's config value is present exactly when `SKIP_CATALOG` is not `true`, so the host and realm-server register the same prefix set; and the default card ids are catalog-derived outside tests and test-realm-owned inside them, so a test's definitions never depend on what a running catalog holds.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/runtime-common/realm-prefixes.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+   */
+  hostConfigKey: string;
+}
+
+export const PREFIX_REALMS: readonly PrefixRealm[] = Object.freeze([
+  {
+    prefix: '@cardstack/base/',
+    alias: 'https://cardstack.com/base/',
+    hostConfigKey: 'resolvedBaseRealmURL',
+  },
```
Comment (habdelra) [inline packages/runtime-common/realm-prefixes.ts:55]: [Claude Code 🤖] Non-blocking, but it is the PR's own thesis applied to itself: both strings in this entry already have a home two files away. `constants.ts` exports `baseRealm` (a `RealmPaths` whose `.url` is `https://cardstack.com/base/`) and `baseRealmRRI` (`@cardstack/base/`) — and `network.ts` read `baseRealm.url` before this change. Writing them again here gives the base realm's spelling three homes.

```ts
import { baseRealm, baseRealmRRI } from './constants.ts';
// …
{
  prefix: baseRealmRRI,
  alias: baseRealm.url,
  hostConfigKey: 'resolvedBaseRealmURL',
},
```

`constants.ts` reaches `index.ts` only through `import type { RealmPermissions }`, which is erased, so importing it here does not create a runtime cycle with the new `export *`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/runtime-common/realm-prefixes.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+   * derivable from the prefix — `@cardstack/openrouter/` is spelled
+   * `resolvedOpenRouterRealmURL`, which no casing rule recovers from the
+   * lowercase prefix segment. The realm-server takes the same URL from
+   * `--toUrl` instead and needs no entry here.
+   *
+   * A property that is absent or empty means the realm is not present in that
+   * build, and the mapping is skipped: the catalog and openrouter realms are
+   * nulled for the host test environment.
+   */
+  hostConfigKey: string;
```
Comment (habdelra) [inline packages/runtime-common/realm-prefixes.ts:47]: [Claude Code 🤖] Regression by omission; non-blocking but in scope. `packages/realm-server/handlers/serve-index.ts` hand-lists the same four keys this field now names — `resolvedBaseRealmURL`, `resolvedCatalogRealmURL`, `resolvedSkillsRealmURL`, `resolvedOpenRouterRealmURL`, each wrapped in `rewriteRealmURL` — so adding a fifth realm still means editing that file too, and the new test only scans launch scripts. A missed entry there ships silently: the deployed host keeps the build-time URL for that realm while the realm-server serves it from a different origin, which is a prefix pointing at the wrong place rather than a prefix that is merely absent.

Since `hostConfigKey` is exactly the name of the property to rewrite, that block can read from the declaration:

```ts
...Object.fromEntries(
  PREFIX_REALMS.map(({ hostConfigKey }) => [
    hostConfigKey,
    rewriteRealmURL(config[hostConfigKey]),
  ]),
),
```

`merge` already ignores `undefined` sources, so a realm absent from the served config stays absent, as today.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/runtime-common/realm-prefixes.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  {
+    prefix: '@cardstack/openrouter/',
+    hostConfigKey: 'resolvedOpenRouterRealmURL',
+  },
+]);
+
+/** Just the prefixes, for comparing one side's registered set against this one. */
+export const PREFIX_REALM_PREFIXES: readonly string[] = Object.freeze(
+  PREFIX_REALMS.map((realm) => realm.prefix),
+);
```
Comment (habdelra) [inline packages/runtime-common/realm-prefixes.ts:73]: [Claude Code 🤖] Follow-up, not this PR. This export is consumed only by the new test, so the check lives entirely in a text scan of shell files. The place it would actually bind is `main.ts` / `worker.ts`, which are where the prefix is registered and where the value is known however it arrived — including through an env var, where a text scan cannot see it (`RESOLVED_SOFTWARE_FACTORY_REALM_URL=https://cardstack.com/software-factory/` would register `@cardstack/software-factory/` and nothing would notice).

There is a second constraint that wants the same assertion. `packages/host/app/lib/trusted-modules.ts` documents it and says it cannot enforce it: no realm may ever be mapped under a name in that file's `hostPackages` set (`boxel-host`, `boxel-ui`, `bxl`, `runtime-common`, …), because such a realm's authored content would be admitted as Host-owned and run uncaged — and it names "the wiring that registers realm mappings" as where the check belongs. `PREFIX_REALMS` is now that wiring. One assertion over the registered `@cardstack/` prefixes — in the declared set, and not in `hostPackages` except for `base` — covers both.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/realm-server/tests/realm-prefix-declaration-test.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+// Every script that launches a realm-server or a worker, in any environment.
+const LAUNCH_SCRIPTS = [
+  'packages/realm-server/scripts/start-production.sh',
+  'packages/realm-server/scripts/start-staging.sh',
+  'packages/realm-server/scripts/start-worker-production.sh',
+  'packages/realm-server/scripts/start-worker-staging.sh',
+  'mise-tasks/services/realm-server',
+  'mise-tasks/services/worker',
+  'mise-tasks/services/worker-test',
+];
```
Comment (habdelra) [inline packages/realm-server/tests/realm-prefix-declaration-test.ts]: [Claude Code 🤖] Non-blocking, but the comment claims a completeness the list does not have. Five other launchers pass `--fromUrl`:

- `mise-tasks/services/realm-server-base`
- `mise-tasks/services/worker-base`
- `mise-tasks/services/test-realms`
- `packages/matrix/support/isolated-realm-server.ts`
- `packages/realm-test-harness/src/isolated-realm-stack.ts`

I ran the extraction over all five: the first four yield only `@cardstack/base/` (plus `@cardstack/skills/` for the matrix one), so adding them to `LAUNCH_SCRIPTS` passes today. The fifth is the interesting one — it spells its arguments `--fromUrl=${publicBaseRealmURL.href}`, unquoted, and the regex here only matches `'…'` or `"…"`, so it extracts nothing. Adding it to the list would make the "no undeclared prefix" test pass over it while seeing zero prefixes.

That vacuum is the part worth fixing regardless of the list: the first test asserts nothing when a script yields no prefixes, so a future change to how these args are written (unquoted, or a variable holding the whole `--fromUrl=…` token) turns the guard off without turning it red. `assert.true(prefixes.length > 0, …)` per scanned script, plus accepting a bare unquoted value in the regex, makes both failures loud.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by habdelra — COMMENTED** [review]
[Claude Code 🤖] Reviewed for whether the declaration actually binds both sides — I traced the prefix derivation in `main.ts` and `worker.ts` against every launcher in the tree, ran the test's extraction over all of them, evaluated `config/environment.js` under each `SKIP_CATALOG` value, and swept the consumers of `resolvedCatalogRealmURL`. I did not review the prerender/definitions-cache read path itself beyond confirming the key-shape argument.

**One blocking-worthy issue: making `resolvedCatalogRealmURL` reachable in a test build changes three more things than `defaultSystemCardId`, and CI cannot see any of them because it sets `SKIP_CATALOG=true`.** The consolidation itself is sound, and the `@cardstack/boxel-ui/` exclusion holds up — it round-trips to the same internal key on both sides, so it is not a divergence.

Recommendations, most to least important:

1. Decide what the test environment should see now that the catalog config value survives — `catalogRealmURL`, the `/_catalog-realms` mock, and `publicRealmURLs` all change under the local default. See the thread on `environment.js` line 237. This extends the thread codex opened; the commit resolves the `defaultSystemCardId` half of it.
2. Normalise `SKIP_CATALOG` to one spelling, or narrow the claim in the new comment — `mise-tasks/services/worker` disagrees with `mise-tasks/services/realm-server` on any value other than `true`. Thread on `environment.js` line 228.
3. Read `serve-index.ts`'s four `resolved*RealmURL` rewrites from `hostConfigKey` — thread on `realm-prefixes.ts`.
4. Make the launch-script scan fail loudly when it extracts nothing, and extend `LAUNCH_SCRIPTS` past the seven listed — thread on the test.
5. Take the base realm's two strings from `constants.ts` rather than re-spelling them.
6. Follow-up, not this PR: assert the registered prefix set at the point of registration, which also gives `trusted-modules.ts` the check it documents and cannot make.

Boxel CLI Tests is red on this head, failing in "Run integration tests against the packed npm install". That job is green on `main`, and I could not reach the failing step's output — the retrievable log tail is entirely the "Print server logs" step. Nothing in the diff reaches that job's code path that I can find: the host assets it consumes are a production build, which this change leaves alone, and `realm-prefixes.ts` resolves fine through runtime-common's `'./*': './*.ts'` export. Worth a re-run before reading anything into it.

Adjacent, out of scope: `trusted-modules.ts`'s doc comment enumerates where realm prefixes get registered ("`network.ts` for the catalog, skills and OpenRouter realms") — accurate before this change, and now one hop out of date for whoever edits that file next.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/host/config/environment.js (hunk 1 above)
Comment (backspace) [reply packages/host/config/environment.js]: [Claude Code 🤖] You're right, and the three sites check out exactly as described — `utils.ts:67`, `routes.ts:359`, `_client.ts:60`. Taken your first option in d80ed3b: don't widen anything but the prefix set.

`resolvedCatalogRealmURL` was answering two questions — *is this realm part of the environment* (realm lists, `/_catalog-realms`, `publicRealmURLs`, default card ids) and *where is it served* (all a prefix mapping needs). `prefixRealmURLs` now carries the second, built before the test environment trims its lists, and the realm-list properties revert to exactly what they were.

Measured across environments:

| environment | `resolvedCatalogRealmURL` | prefix map |
| -- | -- | -- |
| test, `SKIP_CATALOG` unset | undefined (as `main`) | all four |
| test, `SKIP_CATALOG=true` | undefined | catalog absent |
| development | catalog URL | all four |

So the prefix set matches the realm-server under either `SKIP_CATALOG` value, and none of the three sites moves.

On openrouter: it now gets the same treatment rather than a different argument — trimmed from a test's realm lists, prefix registered where the realm is configured. That is a change from `main`, which registered no openrouter prefix in a test build; the justification is the one you'd apply to the catalog, and I'd rather the two be symmetric than have one carry an isolation argument the other doesn't.

The prefixes are spelled in `config/environment.js` because it's CommonJS at build time and can't import `PREFIX_REALMS`. That's a second list, so a host unit test holds its keys to the declaration — same shape as the launch-script scan.

File: packages/host/config/environment.js (hunk 2 above)
Comment (backspace) [reply packages/host/config/environment.js]: [Claude Code 🤖] Both fixed in 48d2e59 and d80ed3b.

`services/worker` now tests `= "true"` like `services/realm-server` and the host config, so `SKIP_CATALOG=1` no longer starts the catalog on the realm-server and registers its prefix on the host while leaving the worker — the process that reads back the cache the prerender writes — without it.

On the comments: you're right, and it's my own project's rule. Both blocks narrated the change instead of stating the contract, and "now" has no anchor for a later reader. They're rewritten as invariants: where each prefix realm is served, versus which realms an environment has; and the catalog owning the default card ids wherever it is part of the environment, which a test build never is.

File: packages/runtime-common/realm-prefixes.ts (hunk 3 above)
Comment (backspace) [reply packages/runtime-common/realm-prefixes.ts:55]: [Claude Code 🤖] Fair — the PR's own thesis applied to itself. Both strings now come from `constants.ts` via `baseRealmRRI` and `baseRealm.url` (d80ed3b), so the base realm's spelling is back to one home. Confirmed the import doesn't create a runtime cycle: `constants.ts` reaches `index.ts` only through an erased `import type`.

File: packages/runtime-common/realm-prefixes.ts (hunk 4 above)
Comment (backspace) [reply packages/runtime-common/realm-prefixes.ts:47]: [Claude Code 🤖] Good catch, and worse than an absent prefix as you say — the host would keep its build-time URL while the realm-server serves that realm elsewhere, so the prefix resolves to the wrong place. Done in 48d2e59.

Both views are generated from the declaration now: the `hostConfigKey`-named properties, and `prefixRealmURLs`, which needs the same rewriting for the deployed host. `hostConfigKey` survived the restructure precisely because this is its consumer — its doc comment now says so, and says it is *not* where the prefix resolves from.

File: packages/runtime-common/realm-prefixes.ts (hunk 5 above)
Comment (backspace) [reply packages/runtime-common/realm-prefixes.ts:73]: [Claude Code 🤖] Agreed, and left as a follow-up. The env-var case is the part a text scan structurally cannot reach, so the scan is a floor rather than the check.

The `trusted-modules.ts` connection is the more interesting half — that file documents a constraint it says it cannot enforce, and names the registration wiring as where it belongs. Worth filing as its own piece of work rather than appending to this PR, since the assertion wants to live at `addRealmMapping`-time in all three processes, not in the declaration.

I did bring that file's doc comment up to date in 2fcc688 — it enumerated `network.ts` registering the catalog, skills and OpenRouter prefixes individually, which this change makes one hop stale.

File: packages/realm-server/tests/realm-prefix-declaration-test.ts (hunk 6 above)
Comment (backspace) [reply packages/realm-server/tests/realm-prefix-declaration-test.ts]: [Claude Code 🤖] Fixed in b27a9228, including the vacuum, which was the right thing to focus on.

All five launchers are in the list, the regex takes a bare unquoted value (so `'--fromUrl=https://…'` in `isolated-realm-stack.ts`, where the whole token is quoted rather than the value, is read), and each script must now yield at least one prefix. Every launcher mounts the base realm, so reading zero means the scan can no longer parse the file rather than a realm being switched off.

Ran the extraction over all twelve before trusting it: they yield between one and four prefixes, none undeclared, and both production scripts carry the whole set. `isolated-realm-stack.ts` yields one — without the bare-value branch it yielded nothing, exactly as you predicted.

**Comment by backspace** [comment]
[Claude Code 🤖] All six addressed; the blocking one restructured the change rather than patching it. Pushed through 2fcc688.

**1 — test environment (blocking).** Took your first option. `resolvedCatalogRealmURL` was answering both *is this realm in the environment* and *where is it served*; `prefixRealmURLs` now carries the second, built before the test environment trims its lists. The realm-list properties revert to `main` exactly, so `utils.ts`, the `/_catalog-realms` mock and `publicRealmURLs` are untouched, while the prefix set still matches the realm-server under either `SKIP_CATALOG` value. Openrouter gets the same treatment rather than a separate argument — noted on the thread as a deliberate change from `main`.

**2 — `SKIP_CATALOG`.** `services/worker` normalised to `= "true"`. And you were right about both new comment blocks describing the change instead of the contract; rewritten as invariants.

**3 — `serve-index`.** Generated from the declaration, both the `hostConfigKey` properties and `prefixRealmURLs`.

**4 — the scan.** Twelve launchers, bare unquoted values accepted, and a per-script floor of one prefix so it fails loudly instead of vacuously. Verified over all twelve.

**5 — `constants.ts`.** Both base strings come from there now.

**6 — assertion at registration.** Left as a follow-up; the `trusted-modules.ts` half makes it worth its own piece of work rather than an appendix here. I did update that file's doc comment, which this change made stale.

On Boxel CLI Tests: it is CS-12695, not this diff. The failing step reports **301 passed, 8 skipped, 0 failed** — the job goes red purely on six uncaught Postgres `57P01` "terminating connection due to administrator command" errors after the tests complete. Same signature on 5931 and 5939 today, and it passed on a later 5939 run without a code change. Your instinct that nothing in the diff reaches it was right.

Thanks — the environment.js finding in particular was the kind I'd verified only halfway: I checked CI wouldn't break, and stopped before asking what happens when the new branch is actually taken.

**Review by habdelra — APPROVED** [review]
